lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -354,6 +355,37 @@ struct WarpOpDeadResult : public OpRewritePattern<WarpSingleLaneOp> {
   }
 };
 
+// If an operand is directly yielded out of the region we can forward it
+// directly and it doesn't need to go through the region.
+struct WarpOpForwardOperand : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Type> resultTypes;
+    SmallVector<Value> yieldValues;
+    auto yield = cast<vector_ext::YieldOp>(
+        warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+    Value valForwarded;
+    unsigned resultIndex;
+    for (OpOperand &operand : yield->getOpOperands()) {
+      auto arg = operand.get().dyn_cast<BlockArgument>();
+      if (!arg || arg.getOwner()->getParentOp() != warpOp.getOperation())
+        continue;
+      Value result = warpOp.getResult(operand.getOperandNumber());
+      Value warpOperand = warpOp.args()[arg.getArgNumber()];
+      if (result.use_empty() || result.getType() != warpOperand.getType())
```

Comment:
Is this something that we support?

```mlir
%0 = vector_ext.warp_execute_on_lane_0(%laneid) args(%v0 : vector<1xf32>) -> (vector<1xf32>) {
 ^bb0(%arg0 : vector<1xf32>) :
  vector_ext.yield %arg0 : vector<1xf32>
}
```

(With the assumption that every lane has the same value.)

Or is this forbidden and we would just capture the `%v0` without going through `args`?

Reply:
That's a good point. I think having warp_execute_on_lane_0 argument and the block argument have the same type should only be legal if the value is uniform. (or maybe we should disallow this completely). I would go with the assumption that this is correct to convert this example to `%v0` but we should think about it more and make sure the semantic is coherent with the rest.